src/storage/data_table.cpp

```diff
@@ -437,7 +437,133 @@ static void VerifyCheckConstraint(TableCatalogEntry &table, Expression &expr, Da
+	}
+
+	// make the data chunk to check
+	vector<LogicalType> types;
+	for (idx_t i = 0; i < table_entry_ptr->columns.size(); i++) {
+		types.emplace_back(table_entry_ptr->columns[i].type);
+	}
+	DataChunk dst_chunk;
+	dst_chunk.InitializeEmpty(types);
+	for (idx_t i = 0; i < src_keys_ptr->size(); i++) {
+		dst_chunk.data[(*dst_keys_ptr)[i]].Reference(chunk.data[(*src_keys_ptr)[i]]);
+	}
+	dst_chunk.SetCardinality(chunk.size());
+	auto data_table = table_entry_ptr->storage.get();
+
+	idx_t count = dst_chunk.size();
+	if (count <= 0) {
+		return;
+	}
+
+	// we need to look at the error messages concurrently in data table's index and transaction local storage's index
+	vector<string> err_msgs, tran_err_msgs;
+	err_msgs.resize(count);
+	tran_err_msgs.resize(count);
+
+	// check whether or not the chunk can be inserted or deleted into the referenced table' storage
+	TableIndexList &table_indices = data_table->info->indexes;
+	table_indices.Scan([&](Index &index) {
+		if (FindColumnIndex(dst_keys_ptr, index.column_ids)) {
+			if (is_append) {
+				index.VerifyAppendForeignKey(dst_chunk, err_msgs.data());
+			} else {
+				index.VerifyDeleteForeignKey(dst_chunk, err_msgs.data());
+			}
+		}
+		return false;
+	});
+
+	// check whether or not the chunk can be inserted or deleted into the referenced table' transaction local storage
+	auto &transaction = Transaction::GetTransaction(context);
+	bool transaction_check = transaction.storage.Find(data_table);
+	if (transaction_check) {
+		vector<unique_ptr<Index>> &transact_index_vec = transaction.storage.GetIndexes(data_table);
+		for (idx_t i = 0; i < transact_index_vec.size(); i++) {
+			if (FindColumnIndex(dst_keys_ptr, transact_index_vec[i]->column_ids)) {
+				if (is_append) {
+					transact_index_vec[i]->VerifyAppendForeignKey(dst_chunk, tran_err_msgs.data());
+				} else {
+					transact_index_vec[i]->VerifyDeleteForeignKey(dst_chunk, tran_err_msgs.data());
+				}
+			}
+		}
+	}
+
+	// we need to look at the error messages concurrently in data table's index and transaction local storage's index
+	for (idx_t i = 0; i < count; i++) {
+		if (!transaction_check) {
+			// if there is no transaction-local data we only need to check if there is an error message in the main
+			// index
+			if (!err_msgs[i].empty()) {
+				throw ConstraintException(err_msgs[i]);
+			} else {
+				continue;
+			}
+		}
+		if (is_append) {
+			// if we are appending we need to check to ensure the foreign key exists in either the transaction-local
+			// storage or the main table
+			if (!err_msgs[i].empty() && !tran_err_msgs[i].empty()) {
+				throw ConstraintException(err_msgs[i]);
+			} else {
+				continue;
+			}
+		}
+		// if we are deleting we need to ensure the foreign key DOES NOT exist in EITHER the transaction-local storage
+		// OR the main table
+		if (!err_msgs[i].empty() || !tran_err_msgs[i].empty()) {
+			string &err_msg = err_msgs[i];
```

Comment:
This is not triggered - and I'm not sure this condition is correct. Could you add a test that triggers this and verify that this is correct?

Reply:
It is triggered in the following case.
CREATE TABLE pkt(i INTEGER PRIMARY KEY);
CREATE TABLE fkt(j INTEGER, FOREIGN KEY (j) REFERENCES pkt(i));
INSERT INTO pkt VALUES (1);
BEGIN TRANSACTION;
INSERT INTO fkt VALUES (1);
DELETE FROM pkt WHERE i=1;

I will add this test.